### PR TITLE
DRBG reseed from parent with lower security strength.

### LINKED ISSUE #7118: DRBG: Using a weaker strength DRBG as a source is not currently supported.
Currently there is only one DRBG implementation (DRBG_ctr). It is only possible to set one strength for the master and slave DRBG's. 
If alternate DRBG's are added (Hash or HMAC) (See PR #6779) then it is possible to add weaker strength DRBG's as the master DRBG.

If additional DRBG's are supported then a decision will need to be made to either 
  (1) Support this feature OR
  (2) Document that it is not supported.

----------------------------------------------------------------------------------------
NOTES:

rand_lib.c contains the following code:

``` C
rand_drbg_get_entropy(....)
{
  .....
    if (drbg->parent && drbg->strength > drbg->parent->strength) {
        /*
         * We currently don't support the algorithm from NIST SP 800-90C
         * 10.1.2 to use a weaker DRBG as source
         */
        RANDerr(RAND_F_RAND_DRBG_GET_ENTROPY, RAND_R_PARENT_STRENGTH_TOO_WEAK);
        return 0;
    }
}
```

### PR DESCRIPTION
**Pull request by paulidale**
This implements the algorithm in NIST SP 800-90C section 10.1.2 _"Accessing a Source DRBG with Prediction Resistance to Obtain any Security Strength"_.  Apart from step 6.  This permits a DRBG to be seeded from a parent DRBG of lower security strength so long as the parent has access to a live entropy source.

The missing step 6 runs the appropriate generating function over the collected seed material if what was collected was too large.  The size limit used is (at least) 2³¹-1 bytes which is significantly more than we'll ever collect even under our worst quality assumptions.

- [ ] documentation is added or updated
- [x] tests are added or updated

Based upon #8647 
Fixes #7118

### CODE REVIEW

File: crypto/rand/rand_lib.c [hunk 1]
```diff
@@ -137,52 +137,76 @@ size_t rand_drbg_get_entropy(RAND_DRBG *drbg,
...
+         * Check if NIST SP 800-90C 10.1.2 is required to seed from a weaker
+         * parent DRBG.  This standard mandates prediction resistance with
+         * a halved strength per collection and the collections concatenated
+         * together.
+         *
+         * The iteration limit is computed using bytes not bits and it uses
+         * number of bytes wanted rather than the parent DRBG strength.  The
+         * former (bytes vs bits) doesn't result in a difference from the
+         * specified algorithm and the latter is conservative because more
+         * bytes are easked for than the DRBG strength -- i.e. more entropy
```
Comment (slontis) [inline crypto/rand/rand_lib.c]: NIT easked

File: doc/man3/RAND_DRBG_reseed.pod [hunk 2]
```diff
@@ -13,7 +13,8 @@ RAND_DRBG_set_reseed_defaults
  #include <openssl/rand_drbg.h>
 
  int RAND_DRBG_reseed(RAND_DRBG *drbg,
-                      const unsigned char *adin, size_t adinlen);
+                      const unsigned char *adin, size_t adinlen,
```
Comment (slontis) [inline doc/man3/RAND_DRBG_reseed.pod]: Is this a public API that you are changing here?

Comment (paulidale) [reply doc/man3/RAND_DRBG_reseed.pod]: No, I'm fixing the documentation for a public API.

File: crypto/rand/rand_lib.c (hunk 1 above)
Comment (paulidale) [reply crypto/rand/rand_lib.c]: Fixed

File: doc/man3/RAND_DRBG_reseed.pod (hunk 2 above)
Comment (mspncp) [reply doc/man3/RAND_DRBG_reseed.pod]: Please create a separate pull request for this fix, because it needs to be backported to 1.1.1

**Comment by kroeckx** [comment]
I think the fixes line is wrong

**Comment by paulidale** [comment]
Yes, the fixes line was wrong. I've changed it to #7118

File: doc/man3/RAND_DRBG_reseed.pod (hunk 2 above)
Comment (paulidale) [reply doc/man3/RAND_DRBG_reseed.pod]: See #8678

**Comment by paulidale** [comment]
Closing this one because it seems unwanted, it's only useful in a few cases and would end up being a crutch for poor designs.

**Pull request closed without merging**